# Laboratorio 6: Codificación de Canal (Segunda Parte)
**Curso:** Teoría de la Información Transmisión de Datos

El presente laboratorio da continuidad al laboratorio 5. Tiene como objetivo profundizar en los conceptos de codificación de canal, específicamente en la transmisión de información a través de un canal borradura (BEC). A partir de simulaciones prácticas, se busca comparar los resultados teóricos de la teoría de la información con los obtenidos experimentalmente mediante la generación, transmisión y recepción de secuencias binarias. En particular, se analizarán códigos de repetición como mecanismos sencillos de protección contra errores, evaluando su tasa de codificación y capacidad de detección y corrección.

## Ejercicio 1: Código de repetición triple

Modifique el código del ejercicio 1 del Laboratorio 5 para un canal con borradura y un código de corrección de errores de repetición triple (3,1).

1. Estime la tasa de error de bit resultante (BER).
2. Discuta si los resultados observados coinciden con los valores teóricos esperados.

In [30]:
# Simulación del código de repetición en un Canal con Borradura (BEC)
import numpy as np

def simulate_BEC_repetition_code(N=1000000, f=0.1, n=3, aleatorizacion = 1 ):
    # Valor para representar una borradura
    BORRADO = -1

    # 1. Fuente: Genera N bits aleatorios (0 o 1)
    X = np.random.randint(0, 2, N)

    # 2. Codificador: Repite cada bit n veces
    X_codificado = np.repeat(X, n)

    # 3. Canal BEC: Aplica borraduras con probabilidad f
    Y_recibido = X_codificado.copy().astype(int)
    # Se genera un patrón de borraduras aleatorio
    patron_borradura = np.random.rand(n * N) < f
    # Se marcan los bits correspondientes como borrados
    Y_recibido[patron_borradura] = BORRADO

    # 4. Decodificador:
    # Se agrupan los bits recibidos en bloques de n
    Y_matriz = Y_recibido.reshape(N, n)
    # Se inicializa el vector de bits decodificados
    X_decodificado = np.zeros(N, dtype=int)

    for i in range(N):
        bloque_recibido = Y_matriz[i]
        # Se filtran las borraduras para ver los bits que sí llegaron
        bits_validos = bloque_recibido[bloque_recibido != BORRADO]

        if len(bits_validos) > 0:
            # Si al menos un bit no fue borrado, la decodificación es exitosa.
            # Todos los bits válidos son iguales al bit original.
            X_decodificado[i] = bits_validos[0]
        else:
            # Si TODOS los bits del bloque fueron borrados, el decodificador falla.
            # Se realiza una suposición aleatoria (se elige 0 o 1 al azar).
            if aleatorizacion == 1:
                X_decodificado[i] = np.random.randint(0, 2)
            else:
                X_decodificado[i] = BORRADO

    # 5. Cálculo del BER (Bit Error Rate)
    # Se compara la secuencia original con la decodificada. Un error ocurre
    # solo cuando la decodificación falló Y la suposición aleatoria fue incorrecta.
    ber_simulado = np.mean(X != X_decodificado)
    print(f"Canal: Binario con Borradura (BEC)                                           ******************")
    print(f"Tasa de Error de Bit (BER) simulada con código ({nk},1) y f={f:.2%} y mensaje de {N} bits: {ber_simulado:.8f}")

    return

def BER_teorico_BEC(f, n, aleatorizacion = 1):
    # Un error de decodificación ocurre si y solo si los n bits repetidos son borrados.
    # La probabilidad de que n bits independientes sean borrados es f^n.
    prob_fallo_decodificacion = f ** n

    # Cuando el decodificador falla, hace una suposición aleatoria.
    # La probabilidad de que esta suposición sea incorrecta es 0.5.
    # El BER es la probabilidad de fallo multiplicada por 0.5.
    if aleatorizacion == 1:
        ber_teorico = 0.5 * prob_fallo_decodificacion
    else:
        ber_teorico = prob_fallo_decodificacion

    print(f"Tasa de Error de Bit (BER) teórica  con código ({nk},1) y f={f:.2%} y mensaje de {N} bits: {ber_teorico:.8f}")
    return

# --- Parámetros de la Simulación ---
f = 0.1           # Probabilidad de borradura (10%)
nk = 3            # Código de repetición (3,1)
N = 100*(10**nk)  # Número de bits a simular para mayor precisión

# --- Ejecución ---
print("*"*120)
print(f"Simulación del código de repetición en un Canal con Borradura (BEC) - Opcion CON ALEATORIZACION si fallan los {nk} bits del mensaje")
simulate_BEC_repetition_code(N, f, nk)
BER_teorico_BEC(f, nk)
print("*"*120)
print(f"Simulación del código de repetición en un Canal con Borradura (BEC) - Opcion SIN ALEATORIZACION si fallan los {nk} bits del mensaje")
simulate_BEC_repetition_code(N, f, nk, -1)
BER_teorico_BEC(f, nk, -1)
print("*"*120)


************************************************************************************************************************
Simulación del código de repetición en un Canal con Borradura (BEC) - Opcion CON ALEATORIZACION si fallan los 3 bits del mensaje
Canal: Binario con Borradura (BEC)                                           ******************
Tasa de Error de Bit (BER) simulada con código (3,1) y f=10.00% y mensaje de 100000 bits: 0.00049000
Tasa de Error de Bit (BER) teórica  con código (3,1) y f=10.00% y mensaje de 100000 bits: 0.00050000
************************************************************************************************************************
Simulación del código de repetición en un Canal con Borradura (BEC) - Opcion SIN ALEATORIZACION si fallan los 3 bits del mensaje
Canal: Binario con Borradura (BEC)                                           ******************
Tasa de Error de Bit (BER) simulada con código (3,1) y f=10.00% y mensaje de 100000 bits: 0.00093000
Tasa 

**Discuta aquí los resultados comparandolos con los resultados teóricos**
En la ejecución realizada se obtuvo:

Tasa de error de bit (BER) simulada con código de repetición (3,1) y
𝑓 = 10.00 %

CON ALEATORIZACION : 0.00049000

SIN ALEATORIZACION : 0.00093000

Tasa de error de bit (BER) teórica con código (3,1) y 𝑓 =10.00 %

CON ALEATORIZACION : 0.00050000

SIN ALEATORIZACION : 0.00100000

Los resultados empíricos coinciden con los valores teóricos, lo que confirma la validez del método de simulación empleado. Este comportamiento es coherente con la teoría, ya que el código de repetición triple reduce significativamente la probabilidad de error. No obstante, esta ganancia en confiabilidad implica un aumento del número de bits transmitidos (aumente x 3), lo que reduce la eficiencia de transmisión.

## Ejercicio 2: Código de de repetición quíntuple

Repita el ejercicio anterior para el código de repetición (5,1).

1. Estime mediante simulación la BER, y compare los resultados con los del ejercicio anterior.
2. Discuta los resultados obtenidos en la simulación y comparelos con los del ejercicio anterior.

In [31]:
# Simulación del código de repetición en un Canal con Borradura (BEC)
import numpy as np

def simulate_BEC_repetition_code(N=1000000, f=0.1, n=3, aleatorizacion = 1 ):
    # Valor para representar una borradura
    BORRADO = -1

    # 1. Fuente: Genera N bits aleatorios (0 o 1)
    X = np.random.randint(0, 2, N)

    # 2. Codificador: Repite cada bit n veces
    X_codificado = np.repeat(X, n)

    # 3. Canal BEC: Aplica borraduras con probabilidad f
    Y_recibido = X_codificado.copy().astype(int)
    # Se genera un patrón de borraduras aleatorio
    patron_borradura = np.random.rand(n * N) < f
    # Se marcan los bits correspondientes como borrados
    Y_recibido[patron_borradura] = BORRADO

    # 4. Decodificador:
    # Se agrupan los bits recibidos en bloques de n
    Y_matriz = Y_recibido.reshape(N, n)
    # Se inicializa el vector de bits decodificados
    X_decodificado = np.zeros(N, dtype=int)

    for i in range(N):
        bloque_recibido = Y_matriz[i]
        # Se filtran las borraduras para ver los bits que sí llegaron
        bits_validos = bloque_recibido[bloque_recibido != BORRADO]

        if len(bits_validos) > 0:
            # Si al menos un bit no fue borrado, la decodificación es exitosa.
            # Todos los bits válidos son iguales al bit original.
            X_decodificado[i] = bits_validos[0]
        else:
            # Si TODOS los bits del bloque fueron borrados, el decodificador falla.
            # Se realiza una suposición aleatoria (se elige 0 o 1 al azar).
            if aleatorizacion == 1:
                X_decodificado[i] = np.random.randint(0, 2)
            else:
                X_decodificado[i] = BORRADO

    # 5. Cálculo del BER (Bit Error Rate)
    # Se compara la secuencia original con la decodificada. Un error ocurre
    # solo cuando la decodificación falló Y la suposición aleatoria fue incorrecta.
    ber_simulado = np.mean(X != X_decodificado)
    print(f"Canal: Binario con Borradura (BEC)                                           ******************")
    print(f"Tasa de Error de Bit (BER) simulada con código ({nk},1) y f={f:.2%} y mensaje de {N} bits: {ber_simulado:.8f}")

    return

def BER_teorico_BEC(f, n, aleatorizacion = 1):
    # Un error de decodificación ocurre si y solo si los n bits repetidos son borrados.
    # La probabilidad de que n bits independientes sean borrados es f^n.
    prob_fallo_decodificacion = f ** n

    # Cuando el decodificador falla, hace una suposición aleatoria.
    # La probabilidad de que esta suposición sea incorrecta es 0.5.
    # El BER es la probabilidad de fallo multiplicada por 0.5.
    if aleatorizacion == 1:
        ber_teorico = 0.5 * prob_fallo_decodificacion
    else:
        ber_teorico = prob_fallo_decodificacion

    print(f"Tasa de Error de Bit (BER) teórica  con código ({nk},1) y f={f:.2%} y mensaje de {N} bits: {ber_teorico:.8f}")
    return

# --- Parámetros de la Simulación ---
f = 0.1           # Probabilidad de borradura (10%)
nk = 5            # Código de repetición (3,1)
N = 100*(10**nk)  # Número de bits a simular para mayor precisión

# --- Ejecución ---
print("*"*120)
print(f"Simulación del código de repetición en un Canal con Borradura (BEC) - Opcion CON ALEATORIZACION si fallan los {nk} bits del mensaje")
simulate_BEC_repetition_code(N, f, nk)
BER_teorico_BEC(f, nk)
print("*"*120)
print(f"Simulación del código de repetición en un Canal con Borradura (BEC) - Opcion SIN ALEATORIZACION si fallan los {nk} bits del mensaje")
simulate_BEC_repetition_code(N, f, nk, -1)
BER_teorico_BEC(f, nk, -1)
print("*"*120)

************************************************************************************************************************
Simulación del código de repetición en un Canal con Borradura (BEC) - Opcion CON ALEATORIZACION si fallan los 5 bits del mensaje
Canal: Binario con Borradura (BEC)                                           ******************
Tasa de Error de Bit (BER) simulada con código (5,1) y f=10.00% y mensaje de 10000000 bits: 0.00000530
Tasa de Error de Bit (BER) teórica  con código (5,1) y f=10.00% y mensaje de 10000000 bits: 0.00000500
************************************************************************************************************************
Simulación del código de repetición en un Canal con Borradura (BEC) - Opcion SIN ALEATORIZACION si fallan los 5 bits del mensaje
Canal: Binario con Borradura (BEC)                                           ******************
Tasa de Error de Bit (BER) simulada con código (5,1) y f=10.00% y mensaje de 10000000 bits: 0.00000980

**Discuta aquí los resultados comparandolos con los resultados teóricos**
En la ejecución realizada se obtuvo:

Tasa de error de bit (BER) simulada con código de repetición (3,1) y
𝑓 = 10.00 %

CON ALEATORIZACION : 0.00000530

SIN ALEATORIZACION : 0.0000980

Tasa de error de bit (BER) teórica con código (5,1) y 𝑓 =10.00 %

CON ALEATORIZACION : 0.00000500

SIN ALEATORIZACION : 0.00001000

Los resultados empíricos coinciden con los valores teóricos, lo que confirma la validez del método de simulación empleado. Este comportamiento es coherente con la teoría, ya que el código de repetición quintuple reduce significativamente la probabilidad de error. No obstante, esta ganancia en confiabilidad implica un aumento del número de bits transmitidos (aumente x 5), lo que reduce la eficiencia de transmisión.